### ISSUE TITLE
DrvFs: compiler reports file missing, but it is present

### ISSUE THREAD

**Issue by Warblefly** [Microsoft/WSL#2712]
On build 17046, while compiling the GCC cross-compiler, it bailed out with a "No such file or directory" error when referencing builtins.def — but the file is present. Could this be a variant of #2448 or #2464?

The error:

```
g++ -fno-PIE -c   -g -O2 -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE   -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings   -DHAVE_CONFIG_H -I. -I. -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/. -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/../include -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/../libcpp/include -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/bld/gcc/./gmp -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gmp -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/bld/gcc/./mpfr/src -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/mpfr/src -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/mpc/src  -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/../libdecnumber -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/../libdecnumber/bid -I../libdecnumber -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/../libbacktrace -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/bld/gcc/./isl/include -I/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/isl/include  -o tree-streamer-in.o -MT tree-streamer-in.o -MMD -MP -MF ./.deps/tree-streamer-in.TPo /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/tree-streamer-in.c
In file included from /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/tree.h:23:0,
                 from /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/tree-streamer.c:27:
/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/tree-core.h:213:24: fatal error: /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/builtins.def: No such file or directory
compilation terminated.
Makefile:1099: recipe for target 'tree-streamer.o' failed
make[1]: *** [tree-streamer.o] Error 1
make[1]: *** Waiting for unfinished jobs....

```
But the file reported as missing _is_ present.

```
ls -l /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/builtins.def
-rwxrwxrwx 1 root root 83663 Dec  1 00:38 /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/builtins.def
```

**Comment by Warblefly** [comment]
After a clean re-start, the compilation has failed again, exactly as above, but with a different file:

`/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/gcc/builtins.def:1001:29: fatal error: chkp-builtins.def: No such file or directory`

But the file is alive and well, where it is supposed to be.

**Comment by Warblefly** [comment]
After a further clean re-start, the same compilation fails in a different place. A file appears to be missing, but it is safely in the right place when found with `ls`:

```
In file included from /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/libgfortran/libgfortran.h:44:0,
                 from /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/src/gcc-7.2.0/libgfortran/generated/pack_c16.c:26:
/mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/bld/gcc/gcc/include/stddef.h:1:15: fatal error: stddef.h: No such file or directory
 #include_next <stddef.h>
               ^~~~~~~~~~
```

but:

```
$ ls -l /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/bld/gcc/gcc/include/stddef.h
-rwxrwxrwx 1 root root 14165 Dec  1 23:08 /mnt/e/Users/john/Documents/ffmpeg/MultimediaTools-mingw-w64/sandbox/bld/gcc/gcc/include/stddef.h
```

**Comment by nutcasev15** [comment]
@Warblefly 
The workaround posted in this link fixes this problem temporarily.
https://github.com/Microsoft/WSL/issues/2448#issuecomment-342254181
It worked for me. I'm on build 16299.64. Need to see if the cumulative patch of 16299.98 fixes this.

**Comment by Warblefly** [comment]
Thank you for replying. I'm afraid this workaround, which I've already tried, is unsuccessful in build 17046. 

It is a slightly different problem: the file is occasionally simply "not found"; this isn't an EINVAL return.

**Comment by therealkenc** [comment]
#2484 (sic).

**Comment by Warblefly** [comment]
Well spotted, @therealkenc.

**Comment by nutcasev15** [comment]
@Warblefly Seems like you got all the compilation issues one can get. Lel
Sorry to hear it doesn't fix your problem. My theory is its a fallback as the command states. It remounts the drive with a previous version of the DrvFS driver. Since I'm on 1709 16299.98, the fallback driver for me is the 1607 driver module. Which works great.

Maybe for you since you are beyond 1709, your fallback is the faulty one present in 16299 because its the current stable one. 

Anyway my issue was this one: https://github.com/Microsoft/WSL/issues/2464#issuecomment-347217184

I just posted in related threads to see if someone knew about it because there was no specific issue for it until now.

@therealkenc The cumulative update to 16299.98 fixes this issue. But it seems performance has taken a hit. Kernel compile went from 240~ with ccache to ~350 with ccache.

**Comment by Warblefly** [comment]
Thanks for the info @nutcasev15 — indeed, whether or not the DrvFs filesystem is mounted with, or without, the "fallback" workaround given some time ago, the fault occurs. It makes the job impossible, unfortunately.

**Comment by AnneTheAgile** [comment]
@Warblefly are you on the latest, per #2448  : " The November 30th, 2017 Cumulative Update (KB4051963) for the Fall Creator's Update contains the fix for this issue. Please install this update, and let us know if you encounter any further issues. To verify if you have installed the update, your OS build number should be 16299.98 or later."

**Comment by Warblefly** [comment]
@AnneTheAgile Thank you for asking these questions. I am happy to offer this information:

(1) The build is 17064.rs_prerelease.171118-1403.
(2) It is difficult to say precisely at this point, but there was an earlier build on which my compilation did work correctly (other than the occasional instance of "fork" not working as reported in #2469)

**Comment by DHowett-MSFT** [comment]
A few members of my team (myself, @rajsesh, @yiyang-msft) are seeing this as well, on builds in the 1705x-1706x range.

**Comment by Warblefly** [comment]
Correction: I mis-typed the build. Here, it's 17046.

**Comment by heldchen** [comment]
this can be "reproduced" fairly "consistently" in my setup in several completely different use cases, see (duplicate) issue @Warblefly linked above.

**Comment by DavidCizek** [comment]
This issue can be reproduced in my environment during GCC crosscompiling (Windows version 17063). Version 17040 works fine.

**Comment by heldchen** [comment]
still an issue in 17074

**Comment by nstrelow** [comment]
Please fix this, it is sooo annoying

**Comment by sunilmut** [comment]
Does anyone has a targeted repro for this issue? I can try with @Warblefly's [project](https://github.com/Warblefly/MultimediaTools-mingw-w64/blob/master/README.md), but compiling that seems to be a long process.

**Comment by heldchen** [comment]
unfortunately the errors (in my cases) only arises when there's lot of file actions going on, i.e. during long compile/render processes.

**Comment by sunilmut** [comment]
Doesn't matter from which angle you look at this issue, it looks like a race condition. It will be a difficult one to chase. But, if someone can provide some specific repro here with any good degree of repro, it will be very helpful.

I tried @Warblefly's project, but, it has lot of dependencies and the instructions in the repo do not cover all of them. Chasing each one of them down seems time consuming. I am trying other projects to see if I can get a local repro.

**Comment by Warblefly** [comment]
Thank you for checking @sunilmut — just updated the repo from my live, functioning tree. I'm now compiling it under Ubuntu under Hyper-V Manager and it is perfect. But under Ubuntu Bash on Windows, the failures at random points continue. I don't test Ubuntu Bash on Windows any more; not enough time.

**Comment by DavidCizek** [comment]
Isssue still exists in 17093. I tried to create simple program, which simulates this behavior. On DrvFS reproduces problem, on VolFS works without errors.
`
#include <stdio.h>
#include <pthread.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

void *thread(void *arg)
{
	long num = (long)arg;
	
	while(1)
	{
		int handle;
		const char *fileName = (num % 2) == 0
				? "/mnt/d/Linux/1/Test.c"		// existing file
				: "/mnt/d/Linux/2/Test1.c";		// nonexisting file in nonexisting directory
		
		handle = open(fileName, O_RDONLY);
		if (handle == -1)
		{
			if ((num % 2) == 0)		// existing file
			{
				printf("Existing file open error %d !!!!\n", errno);
			}
			
			continue;
		}
		
		close(handle);
	}
}

#define THREAD_NUMBER	16

int main(void)
{
	pthread_t t[THREAD_NUMBER];
	long i;
	
	for (i = 0; i < THREAD_NUMBER; ++i)
	{
		if (pthread_create(&t[i], NULL, thread, (void *)i) != 0)
		{
			printf("Thread creation error\n");
			return 0;
		}
	}
	
	while(1)
	{
		sleep(0);
	}
}
`

**Comment by heldchen** [comment]
in my scenarios, the error also happens on VolFS, albeit not as often as on DrvFS

**Comment by HinTak** [comment]
I think It is a problem with make -j X ?

**Comment by SvenGroot** [comment]
We believe we have found the cause of this issue, and are working on a fix. We're trying to get this fix into the upcoming Spring Update as well (probably as an update after release).

**Comment by heldchen** [comment]
that is great news, thanks sven!

**Comment by donaldpipowitch** [comment]
@SvenGroot is this issue a confirmed duplicate of https://github.com/Microsoft/WSL/issues/2780? I just saw your comment. It would be so nice to see a fix for https://github.com/Microsoft/WSL/issues/2780 soon.

**Comment by donaldpipowitch** [comment]
> We're trying to get this fix into the upcoming Spring Update as well (probably as an update after release).

Can you confirm this? :) I'd like to opt-out from the Insiders channel now. (AFAIK you need to do the Insiders version matches the stable version?)

**Comment by heldchen** [comment]
@donaldpipowitch as the fix hasn't yet made it into the skip-ahead builds, you can opt out now to be not worse off. I'm sure a possible backport into the stable builds will be announced or negated here once a fix has made it into a future insider build.

**Comment by donaldpipowitch** [comment]
Wow. Cool, Thank you. Dumb question: will this come to stable in the "April" update or in the next big update in about six months?

**Comment by sunilmut** [comment]
@donaldpipowitch - We will need to bake the fix in the insiders for sometime and get some confirmation that it addresses the issue. Then we can try to backport this to SCU ("April") release.

@heldchen - Just to be clear, this is still happening for you on 17655?

**Comment by heldchen** [comment]
@sunilmut yes, debian on 17655

**Comment by sunilmut** [comment]
Thanks @heldchen. Adding @SvenGroot

**Comment by SvenGroot** [comment]
@heldchen Do you have a repro we can try? It sounds like you're hitting a related, yet somehow different, issue.

**Comment by heldchen** [comment]
@SvenGroot it's basically the magento part from #2780 that still fails, the other cases have been resolved for me in 17655. it did fail very often fail on read errors before 17655, whereas in 17655 the failure seems to be reduced to the file delete action.

creating a simple enough repo isn't easy unfortunately due to it happening in a largeish client's project, and only after roughly 1min of the 3min compilation process which creates around 20k new files: It is happening during the frontend theme compiling where 3 (php) process run in parallel, with one of them hitting the issue when trying to cleanup a lock file created in the same process a few seconds before.

adding strace into the mix unfortunately makes the error go away. probably because the build process takes 3x longer and with that somehow avoids the threading issue?

**Comment by therealkenc** [comment]
Did the issue reported here (specifically David's [test case](https://github.com/Microsoft/WSL/issues/2712#issuecomment-365557807)) exist in Fall Creators 16299, or was it introduced after?

**Comment by heldchen** [comment]
my issues started shortly after 16299, first noticed it in build 17040).

**Comment by SvenGroot** [comment]
Can those who still see this issue try again using build 17677, and report the results here?

Thanks!

**Comment by heldchen** [comment]
I've managed to get a strace by replacing `$result = unlink($file)` with `exec("ls -al $file && strace -o /tmp/unlink.trace -f rm $file", $output, $result);`:

**$output**
```
-rw-rw-r-- 1 MyCustomer MyCustomer 10 May 25 01:50 /mnt/t/MyCompany/Projects/clients/MyCustomer/src/www/var/tmp/alternative-source-css.lock
``` 
(just as proof that the file actually exists before trying to delete it)

**unlink.strace**
```
4027  execve("/bin/rm", ["rm", "/mnt/t/MyCompany/Projects/clients"...], [/* 29 vars */]) = 0
4027  brk(NULL)                         = 0x7fffd4e84000
4027  access("/etc/ld.so.nohwcap", F_OK) = -1 ENOENT (No such file or directory)
4027  mmap(NULL, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f3afb640000
4027  access("/etc/ld.so.preload", R_OK) = -1 ENOENT (No such file or directory)
4027  open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 6
4027  fstat(6, {st_mode=S_IFREG|0644, st_size=51925, ...}) = 0
4027  mmap(NULL, 51925, PROT_READ, MAP_PRIVATE, 6, 0) = 0x7f3afb633000
4027  close(6)                          = 0
4027  access("/etc/ld.so.nohwcap", F_OK) = -1 ENOENT (No such file or directory)
4027  open("/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 6
4027  read(6, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\4\2\0\0\0\0\0"..., 832) = 832
4027  fstat(6, {st_mode=S_IFREG|0755, st_size=1689360, ...}) = 0
4027  mmap(NULL, 3795296, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 6, 0) = 0x7f3afb060000
4027  mprotect(0x7f3afb1f5000, 2097152, PROT_NONE) = 0
4027  mmap(0x7f3afb3f5000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 6, 0x195000) = 0x7f3afb3f5000
4027  mmap(0x7f3afb3fb000, 14688, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f3afb3fb000
4027  close(6)                          = 0
4027  mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f3afb630000
4027  arch_prctl(ARCH_SET_FS, 0x7f3afb630700) = 0
4027  mprotect(0x7f3afb3f5000, 16384, PROT_READ) = 0
4027  mprotect(0x7f3afba0e000, 4096, PROT_READ) = 0
4027  mprotect(0x7f3afb623000, 4096, PROT_READ) = 0
4027  munmap(0x7f3afb633000, 51925)     = 0
4027  brk(NULL)                         = 0x7fffd4e84000
4027  brk(0x7fffd4ea5000)               = 0x7fffd4ea5000
4027  open("/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 6
4027  fstat(6, {st_mode=S_IFREG|0644, st_size=2011712, ...}) = 0
4027  mmap(NULL, 2011712, PROT_READ, MAP_PRIVATE, 6, 0) = 0x7f3afb437000
4027  close(6)                          = 0
4027  ioctl(0, TCGETS, {B38400 opost isig icanon echo ...}) = 0
4027  newfstatat(AT_FDCWD, "/mnt/t/MyCompany/Projects/clients/MyCustomer/src/www/var/tmp/alternative-source-css.lock", {st_mode=S_IFREG|0664, st_size=10, ...}, AT_SYMLINK_NOFOLLOW) = 0
4027  geteuid()                         = 2103
4027  newfstatat(AT_FDCWD, "/mnt/t/MyCompany/Projects/clients/MyCustomer/src/www/var/tmp/alternative-source-css.lock", {st_mode=S_IFREG|0664, st_size=10, ...}, AT_SYMLINK_NOFOLLOW) = 0
4027  faccessat(AT_FDCWD, "/mnt/t/MyCompany/Projects/clients/MyCustomer/src/www/var/tmp/alternative-source-css.lock", W_OK) = 0
4027  unlinkat(AT_FDCWD, "/mnt/t/MyCompany/Projects/clients/MyCustomer/src/www/var/tmp/alternative-source-css.lock", 0) = 0
4027  lseek(0, 0, SEEK_CUR)             = -1 ESPIPE (Illegal seek)
4027  close(0)                          = 0
4027  close(1)                          = 0
4027  close(2)                          = 0
4027  exit_group(0)                     = ?
4027  +++ exited with 0 +++
```

**Comment by therealkenc** [comment]
For the lulz I ran David's test case on 17672 and can confirm that much seems okay now.

**Comment by heldchen** [comment]
@therealkenc as said in a few comments before and shown in the above strace, the issue still exists for `delete` actions when there is lot of I/O and concurrency. the test case does only use `open` actions which I can confirm are fixed for me as well.

**Comment by therealkenc** [comment]
Repetition is not constructive.

**Comment by heldchen** [comment]
just wanted to be helpful by pointing out that the test case only tests one part of the problem, or one problem in a family of problems and that the closure of this issue might have been premature.

but before I start "repeating" myself again: where do you want me to continue with my unlink-while-heavy-concurrency-i/o: new issue? #2780?

**Comment by Brian-Perkins** [comment]
@heldchen - the ESPIPE error you are seeing looks like it is on stdin (fd=0) so is likely unrelated. If you can get a repro where one of the other file system commands fails it would be great if you could share (we can route depending on what the error is, and if there is already an existing issue covering it). Unlink on /mnt/x (DrvFS) has had a few reported issues but has been improving over time.

**Comment by heldchen** [comment]
thanks @Brian-Perkins for confirming the ESPIPE is not relevant. 

turns out @therealkenc was right to close this issue after all. I spent some more hours logging all threaded unlink-actions using strace and comparing them. it looks like the Magento framework using a file-based self-rolled locking mechanism. when one thread acquires a lock. unfortunately they did not account for multiple threads trying to acquire the lock at the same time, so in some rare circumstances more than one thread would think that it owns the lock. in these cases, according to my strace logs the chances are actually quite high that two threads try to release their lock by unlinking the file at virtually the same time. in these cases, PHP's file stats cache seems to momentarily reports outdated info to the losing thread, thus resulting in the "errors" I've been observing.

**Comment by therealkenc** [comment]
I guess this manifests is 17134 (1801 RTM).

**Comment by jowadmax** [comment]
@sunilmut thanks for the fix!
Will this bugfix get backported to 1803?

**Comment by sunilmut** [comment]
@jowadmax - Once we get some kind of confirmation (from the community) that the issue is fully resolved by the fixes, yes, the fix will be considered for backporting to 1803.

**Comment by nkokla** [comment]
@sunilmut - Hello and thanks for the fix ! 
I am also impacted by this issue after1803 update and this is very restrictive on my job. I am not sur I can wait the patch. Do you know if a fresh installation (reinstall Windows) can solve this problem ?

**Comment by pgroke-dt** [comment]
@sunilmut Is there a way to install the insider build without permanently enabling preview builds? I might be able to convince my company to let me install _one_ preview build so I can check if the fix works, but I sure don't want to get preview builds all the time.
(I'm seeing this problem many times during one build job, i.e. I have to re-start the build a dozen times or so until it finally succeeds. So if it's fixed, I should be able to tell with good accuracy.)

**Comment by WillsonHaw** [comment]
Can confirm this issue is fixed for me on build 17692. Would love to have this backported to 1803 so I don't have to leave my work machine on a Insider build.

**Comment by pgroke-dt** [comment]
@jowadmax Thanks for the tip, unfortunately I don't have enough space for another Windows installation in a VM on my system.

**Comment by vinz243** [comment]
Not working, my wsl.conf is correct:

```
$ cat /etc/wsl.conf
[automount]
options = case=off
```

Yet I get random error so WSL is unusable for me...

**Comment by Mike-Horwitz** [comment]
@sunilmut     A backport would be great. My 1803 is frustrating me.

**Comment by tstackhouse** [comment]
@therealkenc Seeing this in 17134 (1803), confirming your suspicion. A backport or workaround would be wonderful as I'm unable/unwilling (requirement of Microsoft Account/non-system local account) to install one of the insider builds.

**Comment by therealkenc** [comment]
Yeah the December [OP](https://github.com/Microsoft/WSL/issues/2712#issue-278387935) reported 17046 (before 17134) so there was no real reason to expect it to work on the April Update. My marking it insidertransient for a few days was just "wrong" on my part; apologies. It was [reported](https://github.com/Microsoft/WSL/issues/2712#issuecomment-389476949) that 16299 is okay, fueling the incorrect assumption. At the time it was somewhat confusing as to what broke when, what was being fixed precisely, and in what release (#2484 still flaps in the wind uncategorized). 

Beyond that, four asks for a backport is now three too many. I don't have any specific insight on the interworkings of the backport process (and, I get the impression it is somewhat opaque to the WSL devs as well and not entirely in their control). But, given that it was fixed in late May, this is mid-August, and the Redstone 5 (1809) release is already starting to be locked down for October... if it were me, I wouldn't place a large bet on a QFE. Who knows. Could happen. Either way, rest assured all the devs get all the issue updates whether you ping them or not.
